SEP 20: Add python version support SEP (#26)
saltstack/salt-enhancement-proposals: 0020-py-version-support.md

scheduled release.

How will this impact the Sodium release?
----------------------------------------
Because Python 3.4 became End Of Life on 2019-03-18, we will only support Python versions >= 3.5

How does this impact all supported Operating Systems
----------------------------------------------------
The details below will include the impact for our supported operating systems included here
and some future operating systems we will be supporting:
http://get.saltstack.com/rs/304-PHQ-615/images/SaltStack-Supported-Operating-Systems.pdf

AIX/Solaris
-----------
AIX/Solaris are Enterprise packages and will use salt-bin for their packaging mechanism,
so they will include the correct Python version.

Windows
-------
Python does not come pre-installed on Windows operating systems. We have always included
the Python binary in the package and will continue to do that going forward. If someone
wants to pip install Salt onto windows they will need to install a supported Python version
first, just as they always have. In the case for the Sodium release this will mean they
will need to install Python >=3.5

> OSX

Reviewer: It's called macOS now. Multiple occurrences.

|     30     |    TBD (possibly june 2020) |      3.7        |
|     31     |    TBD (possibly oct 2020)  |      3.7        |
|     32     |    TBD (possibly may 2021)  |      3.8        |

Based off of these dates and Python's End of Life dates, Fedora will not
be of any concern with this change in policy. Fedora releases fairly regularly
and includes the latest Python version. Both packages and pip installs
will work against supported Fedora versions going forward.

SLES/openSUSE
-------------
The General End of Life dates do conflict with the current Python Version
End of Life dates. Since Suse does the packaging here they have been informed
of this policy change and will include a resolution around this with their packages
when these version conflicts occur.

Other Linux
-----------

|     OS     |    End of Life      | Default Python |
|------------|---------------------|----------------|
|     Rhel 6 |    30 Nov 2020      |      2.6       |
|     Rhel 7 |    30 Jun 2024      |      2.7       |
|     Rhel 8 |    05 May 2029      |      None      |
|     Ubuntu 16.04  |    April 2021    |    2.7/3.5 |
|     Ubuntu 18.04  |    April 2023    |    3.6     |

> |     Ubuntu 20.04  |    2030          |    3.8     |

Reviewer 2: ```suggestion
|     Ubuntu 20.04  |    April 2025    |    3.8     |
```

Source: https://ubuntu.com/about/release-cycle.

         salt/modules/vboxmanage.py, salt/utils/virtualbox.py,
         salt/cloud/clouds/virtualbox.py)

Both of these projects have not had commits added to the project in years. This will need to be documented
for the Sodium release to make it clear that these modules will not work unless the upstream projects add
Python 3 support.

Going forward adding more test coverage will help to uncover these projects before we drop a Python version,
but if we do not have enough test coverage, this exercise should be done any time a Python Version is dropped.
There might be a conflict between a Salt dependency and its Python version support that will impact many users.
If this conflict does occur we may need to delay dropping support for the End of Life Python version until that
dependency is updated to the appropriate Python version.

## Who is Affected

Salt Contributors
-----------------
When a Python Version is dropped any contributions will need to ensure they align with the syntax
of the lowest supported Python version. These contributors can look at Python's Release Notes for
details around what is added into a release or removed. For example for Python 3.6

> https://docs.Python.org/3/whatsnew/3.6.html outlines some of the changes and removals added to this version.

Reviewer: > ensure they work with the lowest supported Python version. These contributors can look at Python's Release Notes for details of what has been removed or deprecated. For example https://docs.Python.org/3/whatsnew/3.6.html#removed lists what was removed in Python 3.6.

Most changes are not syntax. No "alignment" is required, and new features aren't relevant.

I notice that this fails to account for existing code that might break. Unless you require all contributors to maintain their contributions indefinitely?

Reviewer: Are new contributions still allowed to use deprecated features? What's the timeline for updating new deprecations that arise in existing code? Will there be an automated system to capture them (e.g. from CI/integration tests)?

Author: I can just remove the word "syntax" and ensure it states contributions need to align with the lowest supported python version. And yes a contributor would need to align with both removals and additions.

New features are relevant. For example, if we begin to support >=Python 3.6 then a contributor can now use fstrings. Whereas when we support  >=python3.5 we cannot accept contributions with fstrings.

You cannot use deprecated features for your contributions. For example if we start supporting >=Python 3.6 you cannot use this deprecated code anymore:

`traceback.Ignore class and traceback.usage, traceback.modname, traceback.fullmodname, traceback.find_lines_from_code, traceback.find_lines, traceback.find_strings, traceback.find_executable_lines methods were removed from the traceback module. They were undocumented methods deprecated since Python 3.2 and equivalent functionality is available from private methods.`

For updating deprecations/testing I specifically call out when we add new python versions here:

https://github.com/saltstack/salt-enhancement-proposals/pull/26/files#diff-d745c39ab3665fe44caa19ff2a6dffdaR62

So at the beginning of each feature development cycle we need to see what python versions are being not only dropped but added and ensure testing is added around that version which should capture the deprecations.

Reviewer: There's a big difference between "can use fstrings" and "must use fstrings". To "align with additions" means the latter, which IMO is unreasonable.

Author: yeah i agree with you that its not a must for new features, I can lighten up the language there.

Salt User
---------
If a Salt user is using the packages from https://repo.saltstack.com they will not be affected
as the Python binary will be included in the package. If they are pip installing the package and
their operating system Python version is an End of Life Python version they will need to upgrade

> their Python version before being able to pip install Salt.

Reviewer: > operating system Python is not supported by the version of salt being installed

Whether it's EoL now doesn't matter if it wasn't when salt was released.

Author: true. i can clean that up to be more clear

Users creating their own packages
---------------------------------
If a user creates their own packages they will only be impacted if they are packaging for an operating
system that includes an End of Life Python Version. They would need to either create their packages using
salt-bin or include a supported version of Python in their packages.

When will this happen?
----------------------

> This Python Version Support change will go into affect as soon as it is approved, but will first impact

Reviewer: effect

the Sodium Release. The Sodium Release is due to be released in June and will support Python versions >= 3.5.

## Alternatives
[alternatives]: #alternatives

> - Alternative is to keep supporting insecure Python versions that line up with supported operating systems

Reviewer: Supported operating systems will not have an insecure Python. The work that RedHat/Debian/etc. does is to ensure their LTS systems continue to have necessary security patches, even if the upstream developers don't provide them.

Author: not all of our supported operating systems follow this approach. I can change the syntax to say End of Life instead of insecure though to align with all of them.

Reviewer: Which systems does Salt support, that do not support their own core packages?

Author: I could not see similar security support for macOS, aix and solaris. Please correct me if im wrong.

Reviewer: I'd be very surprised if IBM and Oracle are leaving unpatched public vulnerabilities in their supported commercial systems. Looks like Apple have or are shortly going to stop packaging Python themselves.

Author: Yeah I honestly just couldn't find documentation. Either way I updated the phrase to be End of Life here instead of insecure.

- Feature Name: Python Version Support
- Start Date: 2020-03-24
- SEP Status: Draft
- SEP PR: (leave this empty)
- Salt Issue: (leave this empty)

# Summary
[summary]: #summary

This SEP is being created to add a policy around Python versions Salt will support and
detail how we will support Python versions going forward. This policy change is to align Salt Releases
with Python's End of Life schedule.

# Motivation
[motivation]: #motivation

Documentation and Communication
-------------------------------
One of the motivating factors around this policy change is to clarify our Python version support
for each Salt release. If it is clearly defined, then packagers, community members, and Salt users will
know when a Python version will be dropped for a given Salt release. In the past we had never clearly
defined how our Python version support lined up with Python's End of Life schedule. We previously made a
best effort to support the Python versions that lined up with the operating systems Python versions, but never clearly
defined this and did not always follow this in every case. For example, when we dropped Python 2.6 support

> we still supported Redhat 6, which had Python 2.6 installed by default. To continue supporting Redhat 6 while also

Reviewer 2: ```suggestion
we still supported Red Hat 6, which had Python 2.6 installed by default. To continue supporting Red Hat 6 while also
```

removing support for the End of Life Python 2.6 version, we included Python 2.7 in our packages
for that operating system.

Including Python in Packages
----------------------------
Another motivating factor to make this policy change is our shift to including Python in our

> packages. For the Sodium release we will be using salt-bin (https://github.com/saltstack/salt-bin) to

Reviewer 3: Salt-bin is now pop-build https://gitlab.com/saltstack/pop/pop-build/

Reviewer 4: ```suggestion
packages. For the Sodium release we will be using pop-build (https://gitlab.com/saltstack/pop/pop-build/) to
```

package for our Fully Supported Platforms.
(http://get.saltstack.com/rs/304-PHQ-615/images/SaltStack-Supported-Operating-Systems.pdf)
This will allow us to easily continue supporting operating systems that include End of Life Python
versions by default by providing the Python binary in the Salt package included in a Salt release.

Security and Maintenance Concerns
---------------------------------
The last motivating factor surrounds security and maintenance concerns. Continuing to support an End
of Life Python version adds additional maintenance around testing and development for each version.
Supporting these versions also puts users at a security risk, since no more security fixes will be

> submited and released with that Python version if the operating system does not choose to update

Reviewer 2: ```suggestion
submitted and released with that Python version if the operating system does not choose to update
```

it themselves.

# Design
[design]: #detailed-design

The design goal of this SEP is to define a clear policy around Salt's Python version support.
This new policy defines only supporting major Python versions until they reach End of Life. You can see

> the current status of supported Python branches here https://devguide.Python.org/#status-of-Python-branches

Reviewer 2: ```suggestion
the current status of supported Python branches here https://devguide.python.org/#status-of-Python-branches
```

with the associated End Of Life date for each branch. You can also see details about Python versions that

> have already become End Of Life here: https://devguide.Python.org/devcycle/#end-of-life-branches.

Reviewer 2: ```suggestion
have already become End Of Life here: https://devguide.python.org/devcycle/#end-of-life-branches.
```

What if a Python version drops support relatively close to a Salt Release?
--------------------------------------------------------------------------
We may drop support for a Python version if the End of Life date lands relatively close to a Salt release.
For example if a Python version is End of Life a couple weeks before a Salt release, support for that
Python version may drop in that Salt version. We will need to communicate and decide when a Python version
is dropped at the beginning of a development cycle.

What about new Python versions?
-------------------------------
At the beginning of a development cycle for each feature release, not only do we need to align
with the Python versions that will become End of Life, but also new Python versions that will need
to be added. Tests running against the new Python version will need to be added and run on the
master branch. Any issues found will need to have issues created and assigned out for the
scheduled release.

How will this impact the Sodium release?
----------------------------------------
Because Python 3.4 became End Of Life on 2019-03-18, we will only support Python versions >= 3.5

> How does this impact all supported Operating Systems
> ----------------------------------------------------

Reviewer 2: ```suggestion
How does this impact all supported Operating Systems?
-----------------------------------------------------
```

affect macOS.

|     OS     |    End of Life      | Default Python |
|------------|---------------------|----------------|
|   10.13    |    September 2020   |    2.7         |
|   10.14    |    September 2021   |    2.7         |
|   10.15    |    TBD              |    2.7         |
|   10.16    |    TBD              |   no default   |

We have always included the Python binary with the mac packages, so that will not change for
users that use the mac Python packages. Starting with the Sodium release we will no longer
support Python 2.7, so if users install Salt on macOS using pip or any other means they will
need to install Python >= 3.5.

Fedora
------
Fedora releases a new version every 6 months and supports these releases for approximately 13 months.
Our current support mirrors that of the upstream and they are currently supporting Fedora 30, 31 and
32. They do not have current EOL dates for these releases yet here:
https://endoflife.software/operating-systems/linux/fedora

|     OS     |    End of Life              |  Default Python |
|------------|-----------------------------|-----------------|

> |     30     |    TBD (possibly june 2020) |      3.7        |
> |     31     |    TBD (possibly oct 2020)  |      3.7        |
> |     32     |    TBD (possibly may 2021)  |      3.8        |

Reviewer 2: ```suggestion
|     30     |    TBD (possibly June 2020) |      3.7        |
|     31     |    TBD (possibly Oct 2020)  |      3.7        |
|     32     |    TBD (possibly May 2021)  |      3.8        |
```

Based off of these dates and Python's End of Life dates, Fedora will not
be of any concern with this change in policy. Fedora releases fairly regularly
and includes the latest Python version. Both packages and pip installs
will work against supported Fedora versions going forward.

SLES/openSUSE
-------------
The General End of Life dates do conflict with the current Python Version

> End of Life dates. Since Suse does the packaging here they have been informed

Reviewer 2: ```suggestion
End of Life dates. Since SUSE does the packaging here they have been informed
```

of this policy change and will include a resolution around this with their packages
when these version conflicts occur.

> Other Linux
> -----------

Reviewer 2: ```suggestion
Other Linuxes
-------------
```

|     OS     |    End of Life      | Default Python |
|------------|---------------------|----------------|

> |     Rhel 6 |    30 Nov 2020      |      2.6       |
> |     Rhel 7 |    30 Jun 2024      |      2.7       |
> |     Rhel 8 |    05 May 2029      |      None      |

Reviewer 2: ```suggestion
|     RHEL 6 |    30 Nov 2020      |      2.6       |
|     RHEL 7 |    30 Jun 2024      |      2.7       |
|     RHEL 8 |    05 May 2029      |      None      |
```

|     Ubuntu 16.04  |    April 2021    |    2.7/3.5 |
|     Ubuntu 18.04  |    April 2023    |    3.6     |
|     Ubuntu 20.04  |    April 2025    |    3.8     |
|     Debian 8      |    30 June 2020  |  2.7, 3.4  |
|     Debian 9      |    June 2022     |  2.7, 3.5  |
|     Debian 10      |    June 2024    |  2.7, 3.7  |

Based on the chart above all of these Operating Systems will or already have conflicted with
Python's End of Life Support dates. We will need to include Python in our packages for these
Operating Systems especially at any given time when their default Python version is considered
End of Life. This does impact users who would want to pip install Salt. They will need to install
a newer version of Python before they can install Salt.

Other Reasonable-Effort Supported Operating Systems
---------------------------------------------------

Arch
----

> Arch is a rolling release and is consistantly updating their Python version, so this is not

Reviewer 2: ```suggestion
Arch is a rolling release and is consistently updating their Python version, so this is not
```

of any concern for this policy.

FreeBSD
-------
According to this https://wiki.freebsd.org/Python/PortsPolicy FreeBSD currently already aligns
with this policy. They only support Python versions until they are End of Life so there are
no concerns with this operating system.

SmartOS
-------
SmartOS includes the latest Python version so will not be impacted by this change.

What about dependencies?
------------------------
Before dropping an End of Life Python version we should have an idea beforehand if Salt's dependencies
do not support a particular Python version with our testing suite. Currently our tests use the dependencies
in the requirements directory in the main Salt project. This does of course mean that if we are not testing
a module we would not have this information beforehand. We are now requiring tests to be added
for PRs so this will help ensure this is caught for those modules, but there are still some modules that do
not have tests currently.

> For the Sodium Release, I installed the libraries from anything defined with :depends: in the documentation

Reviewer 2: ```suggestion
For the Sodium Release, I installed the libraries from anything defined with `:depends:` in the documentation
```

Both of these projects have not had commits added to the project in years. This will need to be documented
for the Sodium release to make it clear that these modules will not work unless the upstream projects add
Python 3 support.

Going forward adding more test coverage will help to uncover these projects before we drop a Python version,
but if we do not have enough test coverage, this exercise should be done any time a Python Version is dropped.
There might be a conflict between a Salt dependency and its Python version support that will impact many users.
If this conflict does occur we may need to delay dropping support for the End of Life Python version until that
dependency is updated to the appropriate Python version.

## Who is Affected

Salt Contributors
-----------------
When a Python Version is dropped the contributor can now use new features in the lowest supported
Python version. For example, f-strings were added in Python 3.6 so when our Python version support is updated
to using >= Python 3.6 a contributor will be able to use f-strings now.

A contributor will also need to be aware of what is deprecated and removed in all supported Python releases when

> making new contributions. For example, in the Python 3.8 release the `macpath` module was removed,

Reviewer 2: Using the backticks for `macpath` here (which I'm in favour of) -- however, from a consistency standpoint, the other uses of "macpath" in this document are not using backticks, so I believe it should be one or the other.

so a contributor would need to ensure they do not write any new contributions using this module.

A contributor can look at Python's Release Notes for details around what is added into a release or removed.
For example for Python 3.6

> https://docs.Python.org/3/whatsnew/3.6.html outlines some of the changes and removals added to this version.

Reviewer 2: ```suggestion
https://docs.python.org/3/whatsnew/3.6.html outlines some of the changes and removals added to this version.
```
